### PR TITLE
Fix/issue 25 - Created time logs for state changes of experiments

### PR DESCRIPTION
**Pull request by YashilDepani**
In this PR, whenever the state changes for the experiment (say from inactive to enrolling, from enrolling to enrollment complete, etc) we have added the time logs. 
1. When the user changes the state of the experiment, it keeps track of every transition of state in the DataBase.
2. On clicking at the Time Logs in the UI, the modal lists transition from enrolling to enrollment-Complete and vice versa for the experiment.

image

image

image

### CODE REVIEW

File: frontend/projects/abtesting/src/assets/i18n/en.json
Comment (JD2455) [inline frontend/projects/abtesting/src/assets/i18n/en.json:137]: Update key name to _home.view-experiment.enrollment-time-logs.text_ as start from the name doesn't used in the value.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.theme.scss
Comment (JD2455) [inline frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.theme.scss:3]: Update class name to use kabab-case.

**Review by JD2455 — CHANGES_REQUESTED** [review]

File: backend/packages/Upgrade/src/api/repositories/StateTimeLogsRepository.ts
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/repositories/StateTimeLogsRepository.ts:10]: Instead of passing individual parameters create an entity of StateTimeLog and pass that.

File: backend/packages/Upgrade/src/api/services/ExperimentService.ts
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/services/ExperimentService.ts:319]: Is this not throwing lint error?
entityManager is an optional parameter. What if it is null?  `entityManager.getRepository` will throw error

File: backend/packages/Upgrade/src/api/repositories/ExperimentRepository.ts
Comment (YashilDepani) [reply backend/packages/Upgrade/src/api/repositories/ExperimentRepository.ts]: code updated.

File: backend/packages/Upgrade/src/api/repositories/ExperimentRepository.ts
Comment (YashilDepani) [reply backend/packages/Upgrade/src/api/repositories/ExperimentRepository.ts]: Code updated.

File: backend/packages/Upgrade/src/api/repositories/ScheduledJobRepository.ts
Comment (YashilDepani) [reply backend/packages/Upgrade/src/api/repositories/ScheduledJobRepository.ts]: Code updated.

File: backend/packages/Upgrade/src/api/repositories/StateTimeLogsRepository.ts
Comment (YashilDepani) [reply backend/packages/Upgrade/src/api/repositories/StateTimeLogsRepository.ts]: The above function has now been removed as we are directly storing the .save(doc) method.

File: backend/packages/Upgrade/src/api/services/AnalyticsService.ts
Comment (YashilDepani) [reply backend/packages/Upgrade/src/api/services/AnalyticsService.ts:276]: As discussed, this logic is correct. I have added comments for more clarity.

File: backend/packages/Upgrade/src/api/services/ExperimentService.ts
Comment (YashilDepani) [reply backend/packages/Upgrade/src/api/services/ExperimentService.ts:263]: Yes, the updatedState is an array.

File: backend/packages/Upgrade/src/api/services/ExperimentService.ts
Comment (YashilDepani) [reply backend/packages/Upgrade/src/api/services/ExperimentService.ts]: Code updated.

File: backend/packages/Upgrade/src/api/services/ExperimentService.ts
Comment (YashilDepani) [reply backend/packages/Upgrade/src/api/services/ExperimentService.ts]: Code updated.

File: backend/packages/Upgrade/src/api/services/ExperimentService.ts
Comment (YashilDepani) [reply backend/packages/Upgrade/src/api/services/ExperimentService.ts]: Code updated.

File: backend/packages/Upgrade/src/api/services/ScheduledJobService.ts
Comment (YashilDepani) [reply backend/packages/Upgrade/src/api/services/ScheduledJobService.ts]: Code updated.

File: backend/packages/Upgrade/test/integration/Experiment/update/ExperimentStartDate.ts
Comment (YashilDepani) [reply backend/packages/Upgrade/test/integration/Experiment/update/ExperimentStartDate.ts]: Code updated.

File: frontend/projects/abtesting/src/app/core/experiments/store/experiments.model.ts
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/core/experiments/store/experiments.model.ts]: Code updated.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.html
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.html]: Code updated.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.scss
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.scss]: Code updated.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.scss
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.scss]: Code updated.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.scss
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.scss]: Code updated.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.scss
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.scss:43]: Code updated.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.scss
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.scss]: Code updated.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts]: Code updated.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts:35]: The logic was correct and I have added comments as discussed.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts]: Code updated.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts]: Code updated.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts]: Code updated.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts]: The variable renamed as camelCase.

File: frontend/projects/abtesting/src/app/features/dashboard/home/pages/view-experiment/view-experiment.component.html
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/features/dashboard/home/pages/view-experiment/view-experiment.component.html]: Code updated.

File: frontend/projects/abtesting/src/assets/i18n/en.json
Comment (YashilDepani) [reply frontend/projects/abtesting/src/assets/i18n/en.json]: Code updated.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.theme.scss
Comment (YashilDepani) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.theme.scss]: Code updated.

File: backend/packages/Upgrade/src/api/repositories/StateTimeLogsRepository.ts
Comment (YashilDepani) [reply backend/packages/Upgrade/src/api/repositories/StateTimeLogsRepository.ts]: I have made an instance of the stateTimeLog class and initiated the values.

File: backend/packages/Upgrade/src/api/services/ExperimentService.ts
Comment (YashilDepani) [reply backend/packages/Upgrade/src/api/services/ExperimentService.ts]: Yes, need to also handle null value. I have updated the code.

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts
Comment (JD2455) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts:35]: Can you format this line of code. Add space before and after operator.

**Review by VivekFitkariwala — APPROVED** [review]

**Review by JD2455 — APPROVED** [review]

File: frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts
Comment (RidhamShah) [reply frontend/projects/abtesting/src/app/features/dashboard/home/components/modal/state-time-logs/state-time-logs.component.ts]: Code updated

**Review by ppratikcr7 — APPROVED** [review]

**Comment by YashilDepani** [comment]
Don't merge the PR, as there are still few enhancements left.

**Comment by YashilDepani** [comment]
This PR, can now be merged.

**Comment by nirmalpatel** [comment]
@msandbothe @amurphy-cl Before merging this PR, please wait for additional comments from @YashilDepani. There is a non-buggy data inconsistency issue with this PR. Since it not a breaking change and can be fixed later and I can create a new 'extension' issue for that.

**Pull request merged**
